docs/configuration.md
---
description: Telliot tweaks and settings to keep your rig running smoothly.
---

# Configuration reference

## CLI reference

Telliot commands and config file options are as the following:

#### Required Flags <a id="docs-internal-guid-d1a57725-7fff-a753-9236-759dd3f42eed"></a>

* `--config` \(path to your config file.\)

#### Telliot Commands

* `accounts`

```
Usage: telliot accounts

Show accounts

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

* `approve`

```
Usage: telliot approve --from=STRING --to=STRING <amount>

Approve tokens

Arguments:
  <amount>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command
      --from=STRING
      --to=STRING

```

* `balance`

```
Usage: telliot balance <account>

Check the balance of an address

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

* `data`

```
Usage: telliot data

Retrieve data from the contract

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --from=UINT-64             the unix timestamp to use as a starting point
                                 for the data retrieval
      --look-back=2h             how far to lookback

```

* `dataserver`

```
Usage: telliot dataserver

launch only a dataserver instance

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

* `decrypt`

```
Usage: telliot decrypt <file>

Decrypts an ecrypted file and write the decrytped version to disk

Arguments:
  <file>    the file to encrypt

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

* `dispute`

```
Usage: telliot dispute <command>

Perform commands related to disputes

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

Commands:
  dispute new --data-id=INT-64 --timestamp=INT-64 <account>
    start a new dispute

  dispute vote --dispute-id=INT-64 <account>
    vote on a open dispute

  dispute list
    list open disputes

  dispute tally
    tally votes for a dispute ID, need to run unlock fee after that

  dispute tally-list
    list tally for disputes

  dispute execute
    after tallying the votes this command transfers the fee to the reporter or
    reported based on the voting

```

* `dispute execute`

```
Usage: telliot dispute execute

after tallying the votes this command transfers the fee to the reporter or
reported based on the voting

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --no-checks                whether to run some additional checks like
                                 balance, did vote etc. Useful to disable when
                                 running against a test contract
      --gas-price=FLOAT-64       gas max fee to use when running the command
      --dispute-id=INT-64        the dispute id
      --all

```

* `dispute list`

```
Usage: telliot dispute list

list open disputes

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --show-closed              also show executed disputes
      --look-back=120h           how far to lookback, the default only few days
                                 since disputes can be voted only for 2 days.

```

* `dispute new`

```
Usage: telliot dispute new --data-id=INT-64 --timestamp=INT-64 <account>

start a new dispute

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command
      --no-checks                whether to run some additional checks like
                                 balance, did vote etc. Useful to disable when
                                 running against a test contract
      --data-id=INT-64           the request id to dispute
      --timestamp=INT-64         the submitted timestamp to dispute

```

* `dispute tally`

```
Usage: telliot dispute tally

tally votes for a dispute ID, need to run unlock fee after that

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command
      --dispute-id=INT-64        the dispute id
      --all

```

* `dispute tally-list`

```
Usage: telliot dispute tally-list

list tally for disputes

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --look-back=120h           how far to lookback, the default only few days
                                 since disputes can be voted only for 2 days.

```

* `dispute vote`

```
Usage: telliot dispute vote --dispute-id=INT-64 <account>

vote on a open dispute

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command
      --dispute-id=INT-64        the dispute id
      --no-checks                whether to run some additional checks like
                                 balance, did vote etc. Useful to disable when
                                 running against a test contract
      --support                  true or false
      --against                  true or false
      --invalid                  true or false

```

* `encrypt`

```
Usage: telliot encrypt <file>

Encrypts a file to be securely stored on disk and later used only with a
password prompt

Arguments:
  <file>    the file to encrypt

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

* `events`

```
Usage: telliot events --name=STRING --contract-name=STRING

Subscribe to watch logs from the network.

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --look-back=DURATION       how far to look for the initiali qyery
      --name=STRING              the name of the log to watch
      --contract-name=STRING     the name of the contract to watch
      --reorg-wait=3s            how long to wait for removed logs from reorg
                                 events

```

* `install-completions`

```
Usage: telliot install-completions

install shell completions

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --uninstall

```

* `stake`

```
Usage: telliot stake <command>

Perform one of the stake operations

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

Commands:
  stake deposit <account>
    deposit a stake

  stake request <account>
    request to withdraw stake

  stake withdraw <account>
    withdraw stake

  stake status <account>
    show stake status

```

* `stake deposit`

```
Usage: telliot stake deposit <account>

deposit a stake

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command
      --no-checks                whether to run some additional checks like
                                 balance, did vote etc. Useful to disable when
                                 running against a test contract

```

* `stake request`

```
Usage: telliot stake request <account>

request to withdraw stake

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command

```

* `stake status`

```
Usage: telliot stake status <account>

show stake status

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

* `stake withdraw`

```
Usage: telliot stake withdraw <account>

withdraw stake

Arguments:
  <account>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command

```

* `submit`

```
Usage: telliot submit --data-id=INT-64 [<account>]

Make a single manual submit to the oracle contracts

Arguments:
  [<account>]

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --data-id=INT-64           the request id to submit
      --gas-price=FLOAT-64       gas max fee to use when running the command
      --custom-submit
      --skip-confirm             submit without confirming, useful for testing

```

* `transfer`

```
Usage: telliot transfer --from=STRING --to=STRING <amount>

Transfer tokens

Arguments:
  <amount>

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

      --gas-price=FLOAT-64       gas max fee to use when running the command
      --from=STRING
      --to=STRING

```

* `version`

```
Usage: telliot version

Show the CLI version information

Flags:
  -h, --help                     Show context-sensitive help.
      --config=STRING            path to config file
      --config-strict-parsing    whether to return an error when the config
                                 contains unknown fields
      --contract=STRING          hex address of the contract to interract with

```

#### .env file options:


* `ETH_PRIVATE_KEYS` \(required\) - list of private keys separated by `,`

* `NODE_URL` \(required\) - websocket node URL \(e.g [wss://mainnet.infura.io/bbbb](wss://mainnet.infura.io/bbbb) or [wss://localhost:8546](ws://localhost:8546) if own node\)


#### Config file options:
```json
{
	"Aggregator": {
		"LogLevel": "Required:false, Default:"
	},
	"Db": {
		"LogLevel": "Required:false, Default:",
		"Path": "Required:false, Default:db",
		"RemoteHost": "Required:false, Default:",
		"RemotePort": "Required:false, Default:9090",
		"RemoteTimeout": {
			"Duration": "Required:false, Default:5s"
		},
		"Retention": {
			"Duration": "Required:false, Default:1440h0m0s"
		}
	},
	"EnvVars": "Required:false, Default:map[]",
	"EthClient": {
		"LogLevel": "Required:false, Default:"
	},
	"PsrTellor": {
		"MinConfidenceDefault": "Required:false, Default:80",
		"MinConfidencePerSymbol": "Required:false, Default:map[USPCE:100]"
	},
	"TrackerBlocks": {
		"LogLevel": "Required:false, Default:"
	},
	"TrackerIndex": {
		"IndexFile": "Required:false, Default:configs/index.json",
		"Interval": {
			"Duration": "Required:false, Default:1m0s"
		},
		"LogLevel": "Required:false, Default:"
	},
	"Transactor": {
		"Transact": "Required:false, Default:false"
	},
	"Web": {
		"ListenHost": "Required:false, Default:",
		"ListenPort": "Required:false, Default:9090",
		"LogLevel": "Required:false, Default:",
		"ReadTimeout": {
			"Duration": "Required:false, Default:0s"
		}
	},
	"envFile": "Required:false, Default:configs/.env"
}
```
Here are the config defaults in json format:
```json
{
	"Aggregator": {
		"LogLevel": ""
	},
	"Db": {
		"LogLevel": "",
		"Path": "db",
		"RemoteHost": "",
		"RemotePort": 9090,
		"RemoteTimeout": "5s",
		"Retention": "1440h0m0s"
	},
	"EnvVars": null,
	"EthClient": {
		"LogLevel": ""
	},
	"PsrTellor": {
		"MinConfidenceDefault": 80,
		"MinConfidencePerSymbol": {
			"USPCE": 100
		}
	},
	"TrackerBlocks": {
		"LogLevel": ""
	},
	"TrackerIndex": {
		"IndexFile": "configs/index.json",
		"Interval": "1m0s",
		"LogLevel": ""
	},
	"Transactor": {
		"Transact": false
	},
	"Web": {
		"ListenHost": "",
		"ListenPort": 9090,
		"LogLevel": "",
		"ReadTimeout": "0s"
	},
	"envFile": "configs/.env"
}
```
### Log levels
Note the default level is "INFO", so to turn down the number of logs, enter "WARN" or "ERROR".

DEBUG - logs everything in INFO and additional developer logs

INFO - logs most information about the reporting operation

WARN - logs all warnings and errors

ERROR - logs only serious errors


pkg/config/config.go
// Copyright (c) The Cryptorium Authors.
// Licensed under the MIT License.

package config

import (
	"context"
	"encoding/json"
	"io"
	"os"
	"path/filepath"
	"time"

	"github.com/cryptoriums/packages/client"
	"github.com/cryptoriums/packages/private_file"
	"github.com/cryptoriums/telliot/pkg/aggregator"
	"github.com/cryptoriums/telliot/pkg/db"
	"github.com/cryptoriums/telliot/pkg/format"
	psr_tellor "github.com/cryptoriums/telliot/pkg/psr/tellor"
	"github.com/cryptoriums/telliot/pkg/tracker/blocks"
	"github.com/cryptoriums/telliot/pkg/tracker/index"
	"github.com/cryptoriums/telliot/pkg/web"
	"github.com/go-kit/log"
	"github.com/go-kit/log/level"
	"github.com/pkg/errors"
	"golang.org/x/tools/godoc/util"
)

type Transactor struct {
	Transact bool
}

// Config is the top-level configuration that holds configs for all components.
type Config struct {
	Transactor    Transactor
	Web           web.Config
	TrackerIndex  index.Config
	TrackerBlocks blocks.Config
	Aggregator    aggregator.Config
	PsrTellor     psr_tellor.Config
	Db            db.Config
	// EnvFile location that include all private details like private key etc.
	EnvFile string `json:"envFile"`
	// Is is set by expanding the os env vars.
	// Not using os.SetEnv so that it doesn't expose those to someone that has access to the machine.
	// If using os.SetEnv for the private key anyone can see it that has access to the machine.
	EnvVars   map[string]string
	EthClient client.Config
}

var DefaultConfig = Config{
	Web: web.Config{
		ListenHost: "", // Listen on all addresses.
		ListenPort: 9090,
	},
	Db: db.Config{
		RemotePort:    9090,
		Path:          "db",
		RemoteTimeout: format.Duration{Duration: 5 * time.Second},
		Retention:     format.Duration{Duration: 60 * 24 * time.Hour}, // 60 days.
	},
	PsrTellor: psr_tellor.Config{
		MinConfidenceDefault: 80,
		MinConfidencePerSymbol: map[string]float64{
			"USPCE": 100,
		},
	},
	TrackerIndex: index.Config{
		Interval:  format.Duration{Duration: time.Minute},
		IndexFile: "configs/index.json",
	},
	EnvFile: "configs/.env",
}

func LoadConfig(ctx context.Context, logger log.Logger, path string, strictParsing bool) (*Config, error) {
	cfg, err := Parse(logger, string(path), strictParsing)
	if err != nil {
		return nil, errors.Wrap(err, "creating config")
	}

	if err := Validate(cfg); err != nil {
		return nil, errors.Wrap(err, "validate config")
	}

	err = LoadEnvFile(ctx, logger, cfg)
	if err != nil {
		return nil, errors.Wrapf(err, "loading the enf file:%v", cfg.EnvFile)
	}
	return cfg, nil
}

func Validate(cfg *Config) error {
MainLoop:
	for pair := range cfg.PsrTellor.MinConfidencePerSymbol {
		for _, psr := range psr_tellor.Psrs {
			if pair == psr.Pair {
				continue MainLoop
			}
		}
		return errors.Errorf("confidence level for invalid psr pair:%v", pair)
	}
	return nil
}

func Parse(logger log.Logger, path string, strictParsing bool) (*Config, error) {

	cfg := &Config{}

	cfgI, err := DeеpCopy(logger, path, cfg, DefaultConfig, strictParsing)
	if err != nil {
		return nil, errors.Wrap(err, "making a config deep copy")
	}
	cfg = cfgI.(*Config)

	return cfg, err
}

func LoadEnvFile(ctx context.Context, logger log.Logger, cfg *Config) error {
	envFileData, err := os.ReadFile(cfg.EnvFile)
	if err != nil {
		return errors.Wrapf(err, "reading the env file:%v", cfg.EnvFile)
	}
	if !util.IsText(envFileData) {
		level.Info(logger).Log("msg", "env file is encrypted", "path", cfg.EnvFile)
		if os.Getenv("KUBERNETES_SERVICE_HOST") != "" {
			transacting := `<span style="color:grey">disabled</span>`
			if cfg.Transactor.Transact {
				transacting = `<span style="color:red">enabled</span>`
			}

			level.Info(logger).Log("msg", "running inside k8s so will wait for web password decrypt input")
			envFileData = private_file.DecryptWithWebPassword(ctx, logger, "<h2>Transacting is:"+transacting+"</h2>", envFileData, cfg.Web.ListenHost, cfg.Web.ListenPort)
		} else {
			envFileData, err = private_file.DecryptWithPasswordLoop(envFileData)
			if err != nil {
				return errors.Wrap(err, "decrypt input file")
			}
		}
	}

	v, err := private_file.SetEnvVars(envFileData)
	cfg.EnvVars = v

	return err
}

func DeеpCopy(logger log.Logger, path string, cfg, cfgDefault interface{}, strictParsing bool) (interface{}, error) {
	var file *os.File
	var fileLoaded bool
	var err error
	if path == "" { // Config flag not provided so try to load the default location.
		path = filepath.Join("configs", "config.json")
		file, err = os.Open(path)
		if err != nil && !os.IsNotExist(err) {
			return nil, errors.Wrap(err, "open config file")
		}
		if err == nil {
			fileLoaded = true
		}
	} else { // When config file location is provided should error when the file doesn't exist.
		// In this case obviously the user want to use some custom file.
		file, err = os.Open(path)
		if err != nil {
			return nil, errors.Wrap(err, "open config file")
		}
		fileLoaded = true
	}

	// Populate the defaults.
	{
		b, err := json.Marshal(cfgDefault)
		if err != nil {
			return nil, errors.Wrap(err, "marshal default config")
		}

		if err := json.Unmarshal(b, cfg); err != nil {
			return nil, errors.Wrap(err, "copy default config")
		}
	}

	//  When a config file is loaded also overwrite the defaults with the values from the config file.
	if fileLoaded {
		dec := json.NewDecoder(file)
		if strictParsing {
			dec.DisallowUnknownFields()
		}
		for {
			// Override defaults with the custom configs.
			if err := dec.Decode(cfg); err == io.EOF {
				break
			} else if err != nil {
				return nil, errors.Wrap(err, "parse config")
			}

		}
	} else {
		level.Warn(logger).Log("msg", "no config file loaded and using the defaults")
	}

	return cfg, nil
}
